slack/management/commands/fetch_slack_activity.py

```diff
@@ -0,0 +1,314 @@
+        (got_lock,) = cur.fetchone()
+        if not got_lock:
+            raise CommandError(
+                "Could not obtain lock: "
+                "another instance of this command must be running."
+            )
+        try:
+            return fn(*args, **kwargs)
+        finally:
+            cur.execute("SELECT pg_advisory_unlock(%s);", [ID])
+
+    return inner
+
+
+@click.command()
+@click.argument("channels", nargs=-1)
+@click.option(
+    "--debug",
+    is_flag=True,
+    help=(
+        "Store all messages seen to be able to "
+        "detect bugs (uses lots of database space)."
+    ),
+)
+@locked
+def command(channels, debug):
+    """
+    Download slack activity from channels the bot is a member of.
+
+    CHANNELS is an optional list of channel names (without the #) to limit to.
+    If not provided, all channels the bot is a member of will be fetched.
+
+    This is resumable -- it can be interrupted and restarted without losing
+    progress.
+
+    Do not run multiple instances of this command in parallel.
+    """
+
+    channels = set(channels)
+    selected_channels = []
+    if channels:
+        for channel_data in get_my_channels():
+            if channel_data["name"] in channels:
+                selected_channels.append(channel_data)
+                channels.remove(channel_data["name"])
+        if channels:
+            raise click.BadParameter(
+                "Could not find channels %r (maybe the bot isn't a member?)" % channels
+            )
+    else:
+        # materialize this generator so we can iterate multiple times
+        selected_channels.extend(get_my_channels())
+
+    for channel_data in selected_channels:
+        with transaction.atomic():
+            channel, created = Channel.objects.update_or_create(
+                id=channel_data["id"],
+                defaults={
+                    "name": channel_data["name"],
+                    "topic": channel_data["topic"]["value"],
+                    "purpose": channel_data["purpose"]["value"],
+                },
+            )
+            if created:
+                # we don't have any messages for this channel we just created
+                channel.channelupdategap_set.create(
+                    oldest_message_ts=None,
+                    newest_message_ts=None,
+                )
+            elif (
+                channel.last_update_ts
+                and not channel.channelupdategap_set.filter(
+                    newest_message_ts=None
+                ).exists()
+            ):
+                # gap from the most recent fetch till now
+                channel.channelupdategap_set.create(
+                    oldest_message_ts=channel.last_update_ts,
+                    newest_message_ts=None,
+                )
+            else:
+                assert (
+                    channel.channelupdategap_set.exists()
+                ), "We must have SOME gaps, time has passed since the last run!"
```

Comment:
Do we want the celery task to crash in this unexpected scenario, or would logging an error and exiting be sufficient?

Reply:
Hitting this line is only possible if there's a logic error in the code. An assertion failure is essentially logging an error and exiting, with the added benefit of a traceback. It'll be easier to troubleshoot the issue with the added context from a traceback, especially in an error reporting tool like Sentry which will show the values of local variables.

Reply:
A `log.error` also sends a stack trace to Sentry, but this is fine here.